StorageAnalyzer crashes

<!-- Thanks for helping me improve SD Maid! Please make sure to fill out as many info's as possible. More info: https://github.com/d4rken-org/sdmaid-se/wiki/Bugs -->

**Describe the bug**
The StorageAnalyzer tool crashes SD Maid repeatedly.

**To Reproduce**
1. Tap the tool name or the eye icon, regardless.
2. Wait for the task to complete
3. Tap any of the devices, regardless
4.  Message about  repeated crash appears


**Smartphone **
 - fairphone 3
 - android 10
 - not rooted, no magisk, no magic whatsoever
 - SD Maid 0.8 beta0

**Debug log**
Stacktrace caught with Scoop app:

```
FATAL EXCEPTION: DefaultDispatcher-worker-10
Process: eu.darken.sdmse, PID: 27362
eu.darken.sdmse.common.files.ReadException: Can't read from path. <-> /storage/emulated/0
	at eu.darken.sdmse.common.files.local.LocalGateway$lookupFiles$3.invokeSuspend(Unknown Source:473)
	at eu.darken.sdmse.common.files.local.LocalGateway$lookupFiles$3.invoke(Unknown Source:12)
	at eu.darken.sdmse.common.files.local.LocalGateway$runIO$2.invokeSuspend(Unknown Source:32)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(Unknown Source:8)
	at kotlinx.coroutines.DispatchedTask.run(Unknown Source:102)
	at kotlinx.coroutines.internal.LimitedDispatcher.run(Unknown Source:11)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Unknown Source:2)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(Unknown Source:76)
	Suppressed: coil.network.HttpException: [StandaloneCoroutine{Cancelling}@334533c, Dispatchers.Default]
Caused by: java.io.IOException: No matching mode available.
	at eu.darken.sdmse.common.files.local.LocalGateway$lookupFiles$3.invokeSuspend(Unknown Source:176)
	... 7 more
```

--- a/app-common-io/src/main/java/eu/darken/sdmse/common/files/GatewaySwitch.kt
+++ b/app-common-io/src/main/java/eu/darken/sdmse/common/files/GatewaySwitch.kt
@@ -4,12 +4,16 @@ import eu.darken.sdmse.common.coroutine.AppScope
 import eu.darken.sdmse.common.coroutine.DispatcherProvider
 import eu.darken.sdmse.common.debug.logging.logTag
 import eu.darken.sdmse.common.files.local.LocalGateway
+import eu.darken.sdmse.common.files.local.LocalPath
 import eu.darken.sdmse.common.files.saf.SAFGateway
+import eu.darken.sdmse.common.files.saf.SAFPath
 import eu.darken.sdmse.common.sharedresource.SharedResource
 import eu.darken.sdmse.common.sharedresource.adoptChildResource
+import eu.darken.sdmse.common.storage.PathMapper
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.toList
 import kotlinx.coroutines.plus
+import okio.IOException
 import okio.Sink
 import okio.Source
 import java.time.Instant
@@ -19,10 +23,11 @@ import javax.inject.Singleton
 
 @Singleton
 class GatewaySwitch @Inject constructor(
-    private val safGateway: SAFGateway,
-    private val localGateway: LocalGateway,
     @AppScope private val appScope: CoroutineScope,
     dispatcherProvider: DispatcherProvider,
+    private val safGateway: SAFGateway,
+    private val localGateway: LocalGateway,
+    private val mapper: PathMapper,
 ) : APathGateway<APath, APathLookup<APath>, APathLookupExtended<APath>> {
 
     private suspend fun <T : APath, R> useGateway(
@@ -67,8 +72,32 @@ class GatewaySwitch @Inject constructor(
         return useGateway(path) { lookup(path) }
     }
 
+    suspend fun lookup(path: APath, type: Type): APathLookup<APath> {
+        val mapped = path.toTargetType(type)
+        return try {
+            useGateway(mapped) { lookup(mapped) }
+        } catch (e: ReadException) {
+            if (type != Type.AUTO) throw e
+
+            val fallback = path.toAlternative()
+            useGateway(fallback) { lookup(fallback) }
+        }
+    }
+
     override suspend fun lookupFiles(path: APath): Collection<APathLookup<APath>> {
         return useGateway(path) { lookupFiles(path) }
+    }
+
+    suspend fun lookupFiles(path: APath, type: Type): Collection<APathLookup<APath>> {
+        val mapped = path.toTargetType(type)
+        return try {
+            useGateway(mapped) { lookupFiles(mapped) }
+        } catch (e: ReadException) {
+            if (type != Type.AUTO) throw e
+
+            val fallback = path.toAlternative()
+            useGateway(fallback) { lookupFiles(fallback) }
+        }
     }
 
     override suspend fun lookupFilesExtended(path: APath): Collection<APathLookupExtended<APath>> {
@@ -81,6 +110,18 @@ class GatewaySwitch @Inject constructor(
 
     override suspend fun exists(path: APath): Boolean {
         return useGateway(path) { exists(path) }
+    }
+
+    suspend fun exists(path: APath, type: Type): Boolean {
+        val mapped = path.toTargetType(type)
+        return try {
+            useGateway(mapped) { exists(mapped) }
+        } catch (e: ReadException) {
+            if (type != Type.AUTO) throw e
+
+            val fallback = path.toAlternative()
+            useGateway(fallback) { exists(fallback) }
+        }
     }
 
     override suspend fun canWrite(path: APath): Boolean {
@@ -122,6 +163,35 @@ class GatewaySwitch @Inject constructor(
     suspend fun walk(path: APath): Collection<APathLookup<APath>> {
         // TODO use safMapper to change path types if error occurs?
         return path.walk(this).toList()
+    }
+
+    private suspend fun APath.toTargetType(type: Type): APath = when (type) {
+        Type.AUTO -> this
+        Type.CURRENT -> this
+        Type.FORCED_LOCAL -> when (this) {
+            is LocalPath -> this
+            is SAFPath -> mapper.toLocalPath(this) ?: throw IOException("Can't map $this to LOCAL")
+            else -> throw IllegalArgumentException("Can't map $this to $type")
+        }
+
+        Type.FORCED_SAF -> when (this) {
+            is LocalPath -> mapper.toSAFPath(this) ?: throw IOException("Can't map $this to SAF")
+            is SAFPath -> this
+            else -> throw IllegalArgumentException("Can't map $this to $type")
+        }
+    }
+
+    private suspend fun APath.toAlternative(): APath = when (this.pathType) {
+        APath.PathType.LOCAL -> mapper.toSAFPath(this as LocalPath) ?: throw IOException("Can't map $this to SAF")
+        APath.PathType.SAF -> mapper.toLocalPath(this as SAFPath) ?: throw IOException("Can't map $this to LOCAL")
+        APath.PathType.RAW -> throw UnsupportedOperationException("Alternative mapping for RAW not available")
+    }
+
+    enum class Type {
+        CURRENT,
+        FORCED_LOCAL,
+        FORCED_SAF,
+        AUTO
     }
 
     companion object {

--- a/app-common-io/src/main/java/eu/darken/sdmse/common/storage/PathMapper.kt
+++ b/app-common-io/src/main/java/eu/darken/sdmse/common/storage/PathMapper.kt
@@ -20,7 +20,7 @@ import javax.inject.Inject
  * Fuck the SAF, this is grating.
  */
 @Reusable
-class SAFMapper @Inject constructor(
+class PathMapper @Inject constructor(
     private val contentResolver: ContentResolver,
     private val storageManager2: StorageManager2,
 ) {

--- a/app/src/main/java/eu/darken/sdmse/analyzer/core/storage/StorageScanner.kt
+++ b/app/src/main/java/eu/darken/sdmse/analyzer/core/storage/StorageScanner.kt
@@ -18,10 +18,8 @@ import eu.darken.sdmse.common.files.APathLookup
 import eu.darken.sdmse.common.files.FileType
 import eu.darken.sdmse.common.files.GatewaySwitch
 import eu.darken.sdmse.common.files.ReadException
-import eu.darken.sdmse.common.files.exists
 import eu.darken.sdmse.common.files.local.File
 import eu.darken.sdmse.common.files.local.LocalPath
-import eu.darken.sdmse.common.files.lookup
 import eu.darken.sdmse.common.files.lookupFiles
 import eu.darken.sdmse.common.files.walk
 import eu.darken.sdmse.common.flow.throttleLatest
@@ -37,7 +35,6 @@ import eu.darken.sdmse.common.progress.Progress
 import eu.darken.sdmse.common.progress.updateProgressPrimary
 import eu.darken.sdmse.common.progress.updateProgressSecondary
 import eu.darken.sdmse.common.root.RootManager
-import eu.darken.sdmse.common.storage.SAFMapper
 import eu.darken.sdmse.common.storage.StorageManager2
 import eu.darken.sdmse.common.user.UserHandle2
 import eu.darken.sdmse.common.user.UserManager2
@@ -56,7 +53,6 @@ class StorageScanner @Inject constructor(
     private val userManager2: UserManager2,
     private val gatewaySwitch: GatewaySwitch,
     private val fileForensics: FileForensics,
-    private val safMapper: SAFMapper,
 ) : Progress.Host, Progress.Client {
 
     private val progressPub = MutableStateFlow<Progress.Data?>(
@@ -88,9 +84,12 @@ class StorageScanner @Inject constructor(
 
         return gatewaySwitch.useRes {
             fileForensics.useRes {
-                val mediaDir = volume?.directory?.let { LocalPath.build(it) }
+                val mediaDir = volume?.directory
+                    ?.let { LocalPath.build(it) }
+                    ?.let { gatewaySwitch.lookup(it, type = GatewaySwitch.Type.AUTO) }
+
                 val folders = mediaDir
-                    ?.lookupFiles(gatewaySwitch)
+                    ?.let { mediaDir.lookupFiles(gatewaySwitch) }
                     ?.filter { it.name != "Android" }
                     ?.mapNotNull { fileForensics.findOwners(it.lookedUp) }
                     ?.filter { it.areaInfo.type == DataArea.Type.SDCARD }
@@ -103,7 +102,6 @@ class StorageScanner @Inject constructor(
 
                 updateProgressSecondary("Scanning media files")
                 val media = mediaDir
-                    ?.lookup(gatewaySwitch)
                     ?.let { scanForMedia(storage, it) }
                     ?: MediaCategory(storage.id, emptySet())
 
@@ -181,28 +179,22 @@ class StorageScanner @Inject constructor(
         // Android/data/<pkg>
         val dataDirPub = publicPath
             ?.let { LocalPath.build(it.path, "Android", "data", pkg.packageName) }
-            ?.takeIf { it.exists(gatewaySwitch) }
-            ?.let { pkgPubDataDir ->
-                val lookup = try {
-                    pkgPubDataDir.lookup(gatewaySwitch)
-                } catch (e: ReadException) {
-                    null
-                }
+            ?.let { pubData ->
+                try {
+                    val lookup = gatewaySwitch.lookup(pubData, type = GatewaySwitch.Type.AUTO)
 
-                when {
-                    lookup?.fileType == FileType.DIRECTORY && !hasApiLevel(33) -> {
+                    if (lookup.fileType == FileType.DIRECTORY && !hasApiLevel(33)) {
                         val children = try {
-                            pkgPubDataDir.walk(gatewaySwitch).map { ContentItem.fromLookup(it) }.toList()
+                            lookup.walk(gatewaySwitch).map { ContentItem.fromLookup(it) }.toList()
                         } catch (e: ReadException) {
                             emptySet()
                         }
-
                         children.plus(ContentItem.fromLookup(lookup)).toNestedContent().single()
+                    } else {
+                        ContentItem.fromInaccessible(pubData)
                     }
-
-                    else -> {
-                        ContentItem.fromInaccessible(pkgPubDataDir)
-                    }
+                } catch (e: ReadException) {
+                    ContentItem.fromInaccessible(pubData)
                 }
             }
 
@@ -234,11 +226,16 @@ class StorageScanner @Inject constructor(
         // Android/media/<pkg>
         val appMediaGroup = publicPath
             ?.let { LocalPath.build(it.path, "Android", "media", pkg.packageName) }
-            ?.takeIf { it.exists(gatewaySwitch) }
-            ?.let { pubMediaDir ->
-                val lookup = pubMediaDir.lookup(gatewaySwitch)
+            ?.let {
+                try {
+                    gatewaySwitch.lookup(it, type = GatewaySwitch.Type.AUTO)
+                } catch (e: ReadException) {
+                    null
+                }
+            }
+            ?.let { lookup ->
                 val children = if (lookup.fileType == FileType.DIRECTORY) {
-                    pubMediaDir.walk(gatewaySwitch).map { ContentItem.fromLookup(it) }.toList()
+                    lookup.walk(gatewaySwitch).map { ContentItem.fromLookup(it) }.toList()
                 } else {
                     emptySet()
                 }
@@ -258,12 +255,18 @@ class StorageScanner @Inject constructor(
                 val owner = it.getOwner(pkg.id) ?: return@filter false
                 !owner.hasFlag(Marker.Flag.CUSTODIAN) && !owner.hasFlag(Marker.Flag.COMMON)
             }
-            .map { ownerInfo ->
-                consumed.add(ownerInfo)
-                val lookup = ownerInfo.areaInfo.file.lookup(gatewaySwitch)
+            .mapNotNull { ownerInfo ->
+                try {
+                    gatewaySwitch.lookup(ownerInfo.areaInfo.file, type = GatewaySwitch.Type.AUTO).also {
+                        consumed.add(ownerInfo)
+                    }
+                } catch (e: ReadException) {
+                    null
+                }
+            }
+            .map { lookup ->
                 val children = if (lookup.fileType == FileType.DIRECTORY) {
-                    ownerInfo.areaInfo.file.walk(gatewaySwitch).map { ContentItem.fromLookup(it) }.toList()
-                        .toNestedContent()
+                    lookup.walk(gatewaySwitch).map { ContentItem.fromLookup(it) }.toList().toNestedContent()
                 } else {
                     emptySet()
                 }
@@ -294,7 +297,8 @@ class StorageScanner @Inject constructor(
         val topLevelContents: Collection<ContentItem> = topLevelDirs.map { ownerInfo ->
             updateProgressSecondary(ownerInfo.areaInfo.file.userReadablePath)
 
-            val lookup = ownerInfo.areaInfo.file.lookup(gatewaySwitch)
+            val lookup = gatewaySwitch.lookup(ownerInfo.areaInfo.file, type = GatewaySwitch.Type.AUTO)
+
             val children = if (lookup.fileType == FileType.DIRECTORY) {
                 lookup.walk(gatewaySwitch).map { ContentItem.fromLookup(it) }.toList()
             } else {

--- a/app/src/main/java/eu/darken/sdmse/analyzer/core/storage/StorageScannerExtensions.kt
+++ b/app/src/main/java/eu/darken/sdmse/analyzer/core/storage/StorageScannerExtensions.kt
@@ -6,17 +6,30 @@ import java.util.LinkedList
 
 
 internal fun Collection<ContentItem>.toNestedContent(): Collection<ContentItem> {
-    val workList = this.sortedByDescending { it.path.segments.size }.toMutableList()
+    val workList = this.sortedBy { it.path.segments.size }.reversed().toMutableList()
 
     val topLevel = mutableListOf<ContentItem>()
 
     val parentIndexMap = mutableMapOf<Segments, Int>()
 
-    workList.forEach { item ->
-        val parentSegs = item.path.segments.subList(0, item.path.segments.size - 1)
+    for (item in workList) {
+        val childSegs = item.path.segments
+        if (childSegs.isEmpty() || childSegs.singleOrNull() == "") {
+            topLevel.add(item)
+            continue
+        }
+
+        val parentSegs = childSegs.subList(0, childSegs.size - 1)
 
         val parentIndex = parentIndexMap[parentSegs]
-            ?: workList.indexOfFirst { it.path.segments == parentSegs }.also { parentIndexMap[parentSegs] = it }
+            ?: workList.indexOfFirst {
+                val segs = it.path.segments
+                if (parentSegs.isEmpty()) {
+                    segs.isEmpty() || segs.singleOrNull() == ""
+                } else {
+                    segs == parentSegs
+                }
+            }.also { parentIndexMap[parentSegs] = it }
 
         if (parentIndex != -1) {
             val parent = workList[parentIndex]
@@ -43,8 +56,8 @@ internal fun ContentItem.toFlatContent(): Collection<ContentItem> {
 
     children
         .map { it.toFlatContent() }
-        ?.flatten()
-        ?.let { result.addAll(it) }
+        .flatten()
+        .let { result.addAll(it) }
 
     return result
 }

--- a/app/src/main/java/eu/darken/sdmse/common/areas/modules/misc/PortableModule.kt
+++ b/app/src/main/java/eu/darken/sdmse/common/areas/modules/misc/PortableModule.kt
@@ -18,7 +18,7 @@ import eu.darken.sdmse.common.files.GatewaySwitch
 import eu.darken.sdmse.common.files.canRead
 import eu.darken.sdmse.common.files.local.LocalPath
 import eu.darken.sdmse.common.files.lookup
-import eu.darken.sdmse.common.storage.SAFMapper
+import eu.darken.sdmse.common.storage.PathMapper
 import eu.darken.sdmse.common.storage.StorageEnvironment
 import eu.darken.sdmse.common.user.UserManager2
 import javax.inject.Inject
@@ -29,7 +29,7 @@ class PortableModule @Inject constructor(
     private val environment: StorageEnvironment,
     private val userManager2: UserManager2,
     private val gatewaySwitch: GatewaySwitch,
-    private val safMapper: SAFMapper,
+    private val pathMapper: PathMapper,
 ) : DataAreaModule {
 
     override suspend fun firstPass(): Collection<DataArea> {
@@ -43,7 +43,7 @@ class PortableModule @Inject constructor(
 
                 if (readablePath == null) {
                     // TODO we don't request SAF permission for this during setup
-                    val safPath = safMapper.toSAFPath(origPath)
+                    val safPath = pathMapper.toSAFPath(origPath)
                     if (safPath?.canRead(gatewaySwitch) == true) {
                         log(TAG, WARN) { "Switched from $origPath to $safPath" }
                         readablePath = safPath

--- a/app/src/main/java/eu/darken/sdmse/common/areas/modules/pub/PublicDataModule.kt
+++ b/app/src/main/java/eu/darken/sdmse/common/areas/modules/pub/PublicDataModule.kt
@@ -19,13 +19,13 @@ import eu.darken.sdmse.common.files.local.LocalGateway
 import eu.darken.sdmse.common.files.local.LocalPath
 import eu.darken.sdmse.common.files.saf.SAFPath
 import eu.darken.sdmse.common.hasApiLevel
-import eu.darken.sdmse.common.storage.SAFMapper
+import eu.darken.sdmse.common.storage.PathMapper
 import javax.inject.Inject
 
 @Reusable
 class PublicDataModule @Inject constructor(
     private val gatewaySwitch: GatewaySwitch,
-    private val safMapper: SAFMapper,
+    private val pathMapper: PathMapper,
 ) : DataAreaModule {
 
     override suspend fun secondPass(firstPass: Collection<DataArea>): Collection<DataArea> {
@@ -42,7 +42,7 @@ class PublicDataModule @Inject constructor(
                                 // If we have root, we need to convert any SAFPath back
                                 when (val target = parentArea.path) {
                                     is LocalPath -> target
-                                    is SAFPath -> safMapper.toLocalPath(target)
+                                    is SAFPath -> pathMapper.toLocalPath(target)
                                     else -> null
                                 }
                             }
@@ -57,11 +57,11 @@ class PublicDataModule @Inject constructor(
                         // On API30 we can do the direct SAF grant workaround
                         when {
                             localGateway.hasRoot() -> when (target) {
-                                is SAFPath -> safMapper.toLocalPath(target)
+                                is SAFPath -> pathMapper.toLocalPath(target)
                                 else -> target
                             }
                             else -> when (target) {
-                                is LocalPath -> safMapper.toSAFPath(target)
+                                is LocalPath -> pathMapper.toSAFPath(target)
                                 is SAFPath -> target
                                 else -> null
                             }

--- a/app/src/main/java/eu/darken/sdmse/common/areas/modules/pub/PublicObbModule.kt
+++ b/app/src/main/java/eu/darken/sdmse/common/areas/modules/pub/PublicObbModule.kt
@@ -19,13 +19,13 @@ import eu.darken.sdmse.common.files.local.LocalGateway
 import eu.darken.sdmse.common.files.local.LocalPath
 import eu.darken.sdmse.common.files.saf.SAFPath
 import eu.darken.sdmse.common.hasApiLevel
-import eu.darken.sdmse.common.storage.SAFMapper
+import eu.darken.sdmse.common.storage.PathMapper
 import javax.inject.Inject
 
 @Reusable
 class PublicObbModule @Inject constructor(
     private val gatewaySwitch: GatewaySwitch,
-    private val safMapper: SAFMapper,
+    private val pathMapper: PathMapper,
 ) : DataAreaModule {
 
     override suspend fun secondPass(firstPass: Collection<DataArea>): Collection<DataArea> {
@@ -41,7 +41,7 @@ class PublicObbModule @Inject constructor(
                             localGateway.hasRoot() -> {
                                 when (val target = parentArea.path) {
                                     is LocalPath -> target
-                                    is SAFPath -> safMapper.toLocalPath(target)
+                                    is SAFPath -> pathMapper.toLocalPath(target)
                                     else -> null
                                 }
                             }
@@ -53,7 +53,7 @@ class PublicObbModule @Inject constructor(
                     }
                     hasApiLevel(30) -> {
                         when (val target = parentArea.path) {
-                            is LocalPath -> safMapper.toSAFPath(target)
+                            is LocalPath -> pathMapper.toSAFPath(target)
                             is SAFPath -> target
                             else -> null
                         }

--- a/app/src/main/java/eu/darken/sdmse/common/areas/modules/pub/SdcardsModule.kt
+++ b/app/src/main/java/eu/darken/sdmse/common/areas/modules/pub/SdcardsModule.kt
@@ -16,7 +16,7 @@ import eu.darken.sdmse.common.files.local.LocalGateway
 import eu.darken.sdmse.common.files.local.LocalPath
 import eu.darken.sdmse.common.rngString
 import eu.darken.sdmse.common.root.RootManager
-import eu.darken.sdmse.common.storage.SAFMapper
+import eu.darken.sdmse.common.storage.PathMapper
 import eu.darken.sdmse.common.storage.StorageEnvironment
 import eu.darken.sdmse.common.user.UserManager2
 import java.io.IOException
@@ -27,7 +27,7 @@ class SdcardsModule @Inject constructor(
     private val storageEnvironment: StorageEnvironment,
     private val userManager2: UserManager2,
     private val gatewaySwitch: GatewaySwitch,
-    private val safMapper: SAFMapper,
+    private val pathMapper: PathMapper,
     private val rootManager: RootManager,
 ) : DataAreaModule {
 
@@ -123,7 +123,7 @@ class SdcardsModule @Inject constructor(
 
         // SAF
         log(TAG) { "$targetPath wasn't accessible trying SAF mapping..." }
-        safMapper.toSAFPath(targetPath)?.let { safPath ->
+        pathMapper.toSAFPath(targetPath)?.let { safPath ->
             val testFileSaf = safPath.child("$TEST_FILE_PREFIX-saf-$rngString")
             try {
                 require(!testFileSaf.exists(gatewaySwitch)) { "Our 'random' testfile already exists? ($testFileSaf)" }

--- a/app/src/main/java/eu/darken/sdmse/common/debug/DebugCardProvider.kt
+++ b/app/src/main/java/eu/darken/sdmse/common/debug/DebugCardProvider.kt
@@ -16,7 +16,7 @@ import eu.darken.sdmse.common.root.RootSettings
 import eu.darken.sdmse.common.root.service.RootServiceClient
 import eu.darken.sdmse.common.shell.ShellOps
 import eu.darken.sdmse.common.shell.root.ShellOpsCmd
-import eu.darken.sdmse.common.storage.SAFMapper
+import eu.darken.sdmse.common.storage.PathMapper
 import eu.darken.sdmse.common.uix.ViewModel3
 import eu.darken.sdmse.main.ui.dashboard.DashboardFragmentDirections
 import eu.darken.sdmse.main.ui.dashboard.items.DebugCardVH
@@ -25,7 +25,7 @@ import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.combine
 import kotlinx.coroutines.flow.distinctUntilChanged
 import kotlinx.coroutines.withTimeoutOrNull
-import java.util.*
+import java.util.UUID
 import javax.inject.Inject
 
 class DebugCardProvider @Inject constructor(
@@ -39,7 +39,7 @@ class DebugCardProvider @Inject constructor(
     private val pkgOps: PkgOps,
     private val automationController: AutomationController,
     private val gatewaySwitch: GatewaySwitch,
-    private val safMapper: SAFMapper,
+    private val pathMapper: PathMapper,
     private val shellOps: ShellOps,
 ) {
 

--- a/app/src/main/java/eu/darken/sdmse/setup/saf/SAFSetupModule.kt
+++ b/app/src/main/java/eu/darken/sdmse/setup/saf/SAFSetupModule.kt
@@ -27,7 +27,7 @@ import eu.darken.sdmse.common.files.saf.SAFPath
 import eu.darken.sdmse.common.files.saf.matchPermission
 import eu.darken.sdmse.common.hasApiLevel
 import eu.darken.sdmse.common.rngString
-import eu.darken.sdmse.common.storage.SAFMapper
+import eu.darken.sdmse.common.storage.PathMapper
 import eu.darken.sdmse.common.storage.StorageEnvironment
 import eu.darken.sdmse.common.storage.StorageManager2
 import eu.darken.sdmse.setup.SetupModule
@@ -40,7 +40,7 @@ class SAFSetupModule @Inject constructor(
     private val contentResolver: ContentResolver,
     private val storageManager2: StorageManager2,
     private val storageEnvironment: StorageEnvironment,
-    private val safMapper: SAFMapper,
+    private val pathMapper: PathMapper,
     private val dataAreaManager: DataAreaManager,
     private val gatewaySwitch: GatewaySwitch,
 ) : SetupModule {
@@ -71,7 +71,7 @@ class SAFSetupModule @Inject constructor(
                 }
                 .mapNotNull { volume ->
                     val targetPath = volume.directory!!.toLocalPath()
-                    val safPath = safMapper.toSAFPath(targetPath)
+                    val safPath = pathMapper.toSAFPath(targetPath)
                     if (safPath == null) {
                         log(TAG, WARN) { "Can't map $targetPath" }
                         return@mapNotNull null
@@ -132,7 +132,7 @@ class SAFSetupModule @Inject constructor(
                 .flatten()
                 .filter { it.exists(gatewaySwitch) }
                 .mapNotNull { targetPath ->
-                    val safPath = safMapper.toSAFPath(targetPath)
+                    val safPath = pathMapper.toSAFPath(targetPath)
                     if (safPath == null) {
                         log(TAG, WARN) { "Can't map $targetPath" }
                         return@mapNotNull null
@@ -180,7 +180,7 @@ class SAFSetupModule @Inject constructor(
             throw IllegalArgumentException("Wrong path")
         }
 
-        safMapper.takePermission(uri)
+        pathMapper.takePermission(uri)
         dataAreaManager.reload()
         refresh()
     }

--- a/app/src/test/java/eu/darken/sdmse/analyzer/core/StorageScannerExtensionsTest.kt
+++ b/app/src/test/java/eu/darken/sdmse/analyzer/core/StorageScannerExtensionsTest.kt
@@ -12,17 +12,21 @@ import testhelpers.BaseTest
 class StorageScannerExtensionsTest : BaseTest() {
 
     @Test
-    fun `content item children nesting`() {
+    fun `content item children nesting variant 1`() {
+        val item0 = ContentItem.fromInaccessible(LocalPath.build())
         val item1 = ContentItem.fromInaccessible(LocalPath.build("folder1"))
-        val item2A = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2A"))
-        val item2B = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2B"))
-        val item3A = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2A/folder3"))
-        val item3B = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2B/folder3"))
+        val item2A = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2A"))
+        val item2B = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2B"))
+        val item3A = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2A", "folder3"))
+        val item3B = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2B", "folder3"))
 
-        val preNesting = setOf(item1, item2A, item2B, item3A, item3B)
+        val preNesting = setOf(item0, item1, item2A, item2B, item3A, item3B)
         val postNesting = preNesting.toNestedContent()
 
-        val item1Post = postNesting.single()
+        val item0Post = postNesting.single()
+        item0Post.path shouldBe item0.path
+
+        val item1Post = item0Post.children.single()
         item1Post.path shouldBe item1.path
 
         item1Post.children.apply {
@@ -42,14 +46,48 @@ class StorageScannerExtensionsTest : BaseTest() {
         }
     }
 
+    @Test
+    fun `content item children nesting variant 2`() {
+        val item0 = ContentItem.fromInaccessible(LocalPath.build(""))
+        val item1 = ContentItem.fromInaccessible(LocalPath.build("folder1"))
+        val item2A = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2A"))
+        val item2B = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2B"))
+        val item3A = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2A", "folder3"))
+        val item3B = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2B", "folder3"))
+
+        val preNesting = setOf(item0, item1, item2A, item2B, item3A, item3B)
+        val postNesting = preNesting.toNestedContent()
+
+        val item0Post = postNesting.single()
+        item0Post.path shouldBe item0.path
+
+        val item1Post = item0Post.children.single()
+        item1Post.path shouldBe item1.path
+
+        item1Post.children.apply {
+            size shouldBe 2
+
+            single { it.path == item2B.path }
+        }
+
+        item1Post.children.single { it.path == item2A.path }.children.apply {
+            size shouldBe 1
+            single { it.path == item3A.path }
+        }
+
+        item1Post.children.single { it.path == item2B.path }.children.apply {
+            size shouldBe 1
+            single { it.path == item3B.path }
+        }
+    }
 
     @Test
     fun `unnest all children`() {
         val item1 = ContentItem.fromInaccessible(LocalPath.build("folder1"))
-        val item2A = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2A"))
-        val item2B = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2B"))
-        val item3A = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2A/folder3"))
-        val item3B = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2B/folder3"))
+        val item2A = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2A"))
+        val item2B = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2B"))
+        val item3A = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2A", "folder3"))
+        val item3B = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2B", "folder3"))
 
         val preNesting = setOf(item1, item2A, item2B, item3A, item3B)
         val postNesting: Collection<ContentItem> = preNesting.toNestedContent()
@@ -60,16 +98,16 @@ class StorageScannerExtensionsTest : BaseTest() {
     @Test
     fun `find child`() {
         val item1 = ContentItem.fromInaccessible(LocalPath.build("folder1"))
-        val item2A = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2A"))
-        val item2B = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2B"))
-        val item3A = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2A/folder3"))
-        val item3B = ContentItem.fromInaccessible(LocalPath.build("folder1/folder2B/folder3"))
+        val item2A = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2A"))
+        val item2B = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2B"))
+        val item3A = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2A", "folder3"))
+        val item3B = ContentItem.fromInaccessible(LocalPath.build("folder1", "folder2B", "folder3"))
 
         val preNesting = setOf(item1, item2A, item2B, item3A, item3B)
         val postNesting: Collection<ContentItem> = preNesting.toNestedContent()
 
         postNesting.findContent {
-            it.path == LocalPath.build("folder1/folder2B/folder3")
+            it.path == LocalPath.build("folder1", "folder2B", "folder3")
         } shouldBe item3B
     }
 }

--- a/app/src/test/java/eu/darken/sdmse/analyzer/core/StorageScannerExtensionsTest2.kt
+++ b/app/src/test/java/eu/darken/sdmse/analyzer/core/StorageScannerExtensionsTest2.kt
@@ -1,0 +1,120 @@
+package eu.darken.sdmse.analyzer.core
+
+import eu.darken.sdmse.analyzer.core.content.ContentItem
+import eu.darken.sdmse.analyzer.core.storage.findContent
+import eu.darken.sdmse.analyzer.core.storage.toFlatContent
+import eu.darken.sdmse.analyzer.core.storage.toNestedContent
+import eu.darken.sdmse.common.files.saf.SAFPath
+import io.kotest.matchers.shouldBe
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+import testhelpers.BaseTest
+
+@RunWith(RobolectricTestRunner::class)
+@Config(sdk = [29])
+class StorageScannerExtensionsTest2 : BaseTest() {
+
+    private val baseTreeUri = "content://com.android.externalstorage.documents/tree/primary%3A"
+
+    @Test
+    fun `content item children nesting base variant 1`() {
+        val item0 = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri))
+        val item1 = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1"))
+        val item2A = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2A"))
+        val item2B = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2B"))
+        val item3A = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2A", "folder3"))
+        val item3B = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2B", "folder3"))
+
+        val preNesting = setOf(item0, item1, item2A, item2B, item3A, item3B)
+        val postNesting = preNesting.toNestedContent()
+
+        val item0Post = postNesting.single()
+        item0Post.path shouldBe item0.path
+
+        val item1Post = item0Post.children.single()
+        item1Post.path shouldBe item1.path
+
+        item1Post.children.apply {
+            size shouldBe 2
+
+            single { it.path == item2B.path }
+        }
+
+        item1Post.children.single { it.path == item2A.path }.children.apply {
+            size shouldBe 1
+            single { it.path == item3A.path }
+        }
+
+        item1Post.children.single { it.path == item2B.path }.children.apply {
+            size shouldBe 1
+            single { it.path == item3B.path }
+        }
+    }
+
+    @Test
+    fun `content item children nesting base variant 2`() {
+        val item0 = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, ""))
+        val item1 = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1"))
+        val item2A = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2A"))
+        val item2B = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2B"))
+        val item3A = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2A", "folder3"))
+        val item3B = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2B", "folder3"))
+
+        val preNesting = setOf(item0, item1, item2A, item2B, item3A, item3B)
+        val postNesting = preNesting.toNestedContent()
+
+        val item0Post = postNesting.single()
+        item0Post.path shouldBe item0.path
+
+        val item1Post = item0Post.children.single()
+        item1Post.path shouldBe item1.path
+
+        item1Post.children.apply {
+            size shouldBe 2
+
+            single { it.path == item2B.path }
+        }
+
+        item1Post.children.single { it.path == item2A.path }.children.apply {
+            size shouldBe 1
+            single { it.path == item3A.path }
+        }
+
+        item1Post.children.single { it.path == item2B.path }.children.apply {
+            size shouldBe 1
+            single { it.path == item3B.path }
+        }
+    }
+
+    @Test
+    fun `unnest all children`() {
+        val item1 = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1"))
+        val item2A = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2A"))
+        val item2B = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2B"))
+        val item3A = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2A", "folder3"))
+        val item3B = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2B", "folder3"))
+
+        val preNesting = setOf(item1, item2A, item2B, item3A, item3B)
+        val postNesting: Collection<ContentItem> = preNesting.toNestedContent()
+
+        postNesting.toFlatContent().toSet() shouldBe preNesting
+    }
+
+    @Test
+    fun `find child`() {
+        val item1 = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1"))
+        val item2A = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2A"))
+        val item2B = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2B"))
+        val item3A = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2A", "folder3"))
+        val item3B = ContentItem.fromInaccessible(SAFPath.build(baseTreeUri, "folder1", "folder2B", "folder3"))
+
+        val preNesting = setOf(item1, item2A, item2B, item3A, item3B)
+        val postNesting: Collection<ContentItem> = preNesting.toNestedContent()
+
+        postNesting.findContent {
+            it.path == SAFPath.build(baseTreeUri, "folder1", "folder2B", "folder3")
+        } shouldBe item3B
+    }
+}

--- a/app/src/test/java/eu/darken/sdmse/analyzer/core/StorageScannerExtensionsTest2.kt
+++ b/app/src/test/java/eu/darken/sdmse/analyzer/core/StorageScannerExtensionsTest2.kt
@@ -13,7 +13,7 @@ import org.robolectric.annotation.Config
 import testhelpers.BaseTest
 
 @RunWith(RobolectricTestRunner::class)
-@Config(sdk = [29])
+@Config(sdk = [29], application = TestApp::class)
 class StorageScannerExtensionsTest2 : BaseTest() {
 
     private val baseTreeUri = "content://com.android.externalstorage.documents/tree/primary%3A"

--- a/app/src/test/java/eu/darken/sdmse/analyzer/core/TestApp.kt
+++ b/app/src/test/java/eu/darken/sdmse/analyzer/core/TestApp.kt
@@ -1,0 +1,5 @@
+package eu.darken.sdmse.analyzer.core
+
+import android.app.Application
+
+class TestApp : Application()

